File: app/shared-dotnet/HyperWhisper.Application/DefaultModePolicy.cs
using HyperWhisper.Data.Entities;
using HyperWhisper.SharedCore;

namespace HyperWhisper.PortableApplication.Persistence;

/// <summary>
/// The default-mode invariant, over the .NET heads' <see cref="Mode"/> entity:
/// exactly one mode carries <see cref="Mode.IsDefault"/>, and that mode's name
/// is fixed (issue #536).
/// </summary>
/// <remarks>
/// <para>
/// The DECISION lives in the shared Rust core (<c>hw-modes</c>), because the
/// tie-break — which mode is promoted when a restored backup leaves none
/// flagged, or two — is a cross-platform contract: three heads that chose
/// differently would restore the same backup to a different default mode each.
/// This class is the .NET half of applying it, shared by the Windows head and
/// the portable (Linux) head, which compile the same <see cref="Mode"/> entity.
/// macOS applies the same core decision from Swift.
/// </para>
/// <para>
/// It deliberately does NOT touch a database. Both heads' stores hand out
/// tracked entities, so the caller mutates them and saves inside its own
/// transaction; a helper that opened its own context could not be part of that
/// transaction and would be a second, racing writer.
/// </para>
/// </remarks>
public static class DefaultModePolicy
{
    /// <summary>
    /// Make exactly one of <paramref name="modes"/> the default, in place.
    /// </summary>
    /// <param name="modes">
    /// EVERY mode that will exist after the write, in display order. Pass the
    /// whole set: "exactly one" cannot be decided from a subset.
    /// </param>
    /// <param name="preferred">
    /// The mode the caller is trying to make the default, or <c>null</c> when it
    /// is only repairing a set that arrived broken.
    /// </param>
    /// <returns>
    /// Whether anything changed. <c>false</c> means the set already satisfied
    /// the invariant, so the caller can skip its save and the change
    /// notification behind it.
    /// </returns>
    public static bool Apply(IReadOnlyList<Mode> modes, Guid? preferred = null)
    {
        ArgumentNullException.ThrowIfNull(modes);
        var plan = SharedCoreBridge.PlanDefaultMode(Flags(modes), preferred);
        if (!plan.Changed) return false;

        foreach (var id in plan.ClearIds)
        {
            var row = modes.FirstOrDefault(mode => mode.Id == id);
            if (row != null) row.IsDefault = false;
        }
        if (plan.DefaultId is { } winner)
        {
            var row = modes.FirstOrDefault(mode => mode.Id == winner);
            if (row != null) row.IsDefault = true;
        }
        return true;
    }

    /// <summary>
    /// Whether <paramref name="stored"/> may be renamed to
    /// <paramref name="newName"/>. The default mode's name is fixed; every other
    /// mode renames freely.
    /// </summary>
    public static PortableModeNameChange CheckRename(Mode stored, string newName)
    {
        ArgumentNullException.ThrowIfNull(stored);
        ArgumentNullException.ThrowIfNull(newName);
        return SharedCoreBridge.CheckModeNameChange(stored.IsDefault, stored.Name, newName);
    }

    /// <summary>
    /// Whether <paramref name="id"/> may have its default flag written to
    /// <paramref name="requestedIsDefault"/>. <paramref name="modes"/> is the
    /// set as it stands BEFORE the write. Only one combination is refused:
    /// clearing the flag on the one mode that carries it.
    /// </summary>
    public static PortableDefaultFlagChange CheckDefaultFlag(
        IReadOnlyList<Mode> modes,
        Guid id,
        bool requestedIsDefault)
    {
        ArgumentNullException.ThrowIfNull(modes);
        return SharedCoreBridge.CheckDefaultModeFlag(Flags(modes), id, requestedIsDefault);
    }

    private static IEnumerable<PortableModeFlags> Flags(IEnumerable<Mode> modes) =>
        modes.Select(mode => new PortableModeFlags(mode.Id, mode.IsDefault, mode.SortOrder));
}

File: app/macos/hyperwhisper/CoreData/DefaultModePolicy.swift
//
//  DefaultModePolicy.swift
//  hyperwhisper
//
//  The default-mode invariant, over Core Data `Mode` objects (issue #536).
//

import CoreData
import Foundation

/// > Exactly one mode carries `isDefault`, and that mode's name is fixed.
///
/// Both halves used to be enforced only by the mode editors — each head
/// disabled the Name field for the mode it happened to be showing, and nothing
/// below the UI checked either one. `PATCH /modes` could therefore rename the
/// default, or set the flag on a second mode, and the editor would then show a
/// disabled field reading the new name directly above the caption PR #535 added
/// saying the name cannot be changed.
///
/// The DECISION is in the shared Rust core (`hw-modes`), not here. Applying it
/// is per-head — this one mutates a Core Data context — but the choice of WHICH
/// mode is promoted when a restored backup left none flagged, or two, is a
/// cross-platform contract: three heads that chose differently would restore the
/// same backup to a different default mode each. `DefaultModePolicy` on the .NET
/// side calls the same core functions.
enum DefaultModePolicy {

    /// Make exactly one of `modes` the default, in place.
    ///
    /// - Parameters:
    ///   - modes: EVERY mode that will exist after the write, in display order.
    ///     "Exactly one" cannot be decided from a subset, and the caller's order
    ///     is the tie-break between equal sort orders.
    ///   - preferred: the mode the caller is trying to make the default, or
    ///     `nil` when it is only repairing a set that arrived broken.
    /// - Returns: whether anything changed. `false` means the set already
    ///   satisfied the invariant, so the caller can skip its save.
    @discardableResult
    static func apply(to modes: [Mode], preferred: UUID? = nil) -> Bool {
        let plan = modePlanDefault(rows: flags(modes), preferred: preferred?.uuidString)
        guard plan.changed else { return false }

        let cleared = Set(plan.clearIds.map { $0.lowercased() })
        for mode in modes where !mode.isDeleted {
            guard let key = mode.id?.uuidString.lowercased() else { continue }
            if cleared.contains(key) {
                mode.isDefault = false
            }
            if let winner = plan.defaultId?.lowercased(), key == winner {
                mode.isDefault = true
            }
        }
        return true
    }

    /// Whether `mode` may be renamed to `newName`. The default mode's name is
    /// fixed; every other mode renames freely.
    static func canRename(_ mode: Mode, to newName: String) -> Bool {
        modeCheckNameChange(
            isDefault: mode.isDefault,
            storedName: mode.name ?? "",
            newName: newName
        ) == .allowed
    }

    /// Whether `id` may have its default flag written to `requested`. `modes` is
    /// the set as it stands BEFORE the write. Only one combination is refused:
    /// clearing the flag on the one mode that carries it.
    static func canWriteDefaultFlag(_ modes: [Mode], id: UUID, requested: Bool) -> Bool {
        modeCheckDefaultFlag(
            rows: flags(modes),
            id: id.uuidString,
            requestedIsDefault: requested
        ) == .allowed
    }

    /// A deleted-but-unsaved object is not part of the set any more, and reading
    /// its properties is a Core Data use-after-delete, so it never reaches the
    /// core.
    private static func flags(_ modes: [Mode]) -> [HwModeFlags] {
        modes.compactMap { mode in
            guard !mode.isDeleted, let id = mode.id else { return nil }
            return HwModeFlags(
                id: id.uuidString,
                isDefault: mode.isDefault,
                sortOrder: Int32(mode.sortOrder)
            )
        }
    }
}
